File: src/session/handshake.rs
use std::net::SocketAddr;
use std::time::{Duration, Instant};

use crate::protocol::constants::*;
use crate::protocol::types::{AudioCodec, PacketType, StreamConfig};
use crate::protocol::Packet;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HandshakeRole {
    Initiator,
    Responder,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HandshakeState {
    Idle,
    WaitingForResponse,
    WaitingForRequest,
    Completed,
    Failed,
}

pub struct Handshake {
    pub role: HandshakeRole,
    pub state: HandshakeState,
    pub remote_addr: Option<SocketAddr>,
    pub ssrc: u32,
    pub config: StreamConfig,
    pub started_at: Option<Instant>,
    pub timeout: Duration,
}

impl Handshake {
    pub fn new(role: HandshakeRole, ssrc: u32, config: StreamConfig) -> Self {
        Self {
            role,
            state: HandshakeState::Idle,
            remote_addr: None,
            ssrc,
            config,
            started_at: None,
            timeout: Duration::from_millis(DEFAULT_HANDSHAKE_TIMEOUT_MS),
        }
    }

    pub fn build_request(&self, seq: u32, ts: u32) -> Packet {
        let mut payload = Vec::with_capacity(9);
        payload.extend_from_slice(&self.ssrc.to_be_bytes());
        payload.extend_from_slice(&self.config.sample_rate.to_be_bytes());
        payload.extend_from_slice(&self.config.bitrate.to_be_bytes());
        payload.push(self.config.codec.to_u8());
        payload.push(self.config.channels as u8);
        Packet::new(PacketType::HandshakeRequest, seq, ts, payload)
    }

    pub fn build_response(&self, seq: u32, ts: u32, accepted: bool) -> Packet {
        let mut payload = vec![if accepted { 0x01 } else { 0x00 }];
        payload.extend_from_slice(&self.ssrc.to_be_bytes());
        Packet::new(PacketType::HandshakeResponse, seq, ts, payload)
    }

    pub fn parse_request(packet: &Packet) -> Option<(u32, StreamConfig)> {
        let payload = &packet.payload;
        if payload.len() < 12 {
            return None;
        }
        let mut ssrc_bytes = [0u8; 4];
        ssrc_bytes.copy_from_slice(&payload[0..4]);
        let remote_ssrc = u32::from_be_bytes(ssrc_bytes);

        let mut sr_bytes = [0u8; 4];
        sr_bytes.copy_from_slice(&payload[4..8]);
        let sample_rate = u32::from_be_bytes(sr_bytes);

        let mut br_bytes = [0u8; 4];
        br_bytes.copy_from_slice(&payload[8..12]);
        let bitrate = u32::from_be_bytes(br_bytes);

        let codec = AudioCodec::from_u8(*payload.get(12).unwrap_or(&0)).unwrap_or(AudioCodec::Opus);
        let channels = *payload.get(13).unwrap_or(&2) as u16;

        Some((
            remote_ssrc,
            StreamConfig {
                sample_rate,
                channels,
                bitrate,
                codec,
                ..Default::default()
            },
        ))
    }

    pub fn parse_response(packet: &Packet) -> Option<(bool, u32)> {
        if packet.payload.len() < 5 {
            return None;
        }
        let accepted = packet.payload[0] == 0x01;

        let mut ssrc_bytes = [0u8; 4];
        ssrc_bytes.copy_from_slice(&packet.payload[1..5]);
        let ssrc = u32::from_be_bytes(ssrc_bytes);
        Some((accepted, ssrc))
    }

    pub fn start(&mut self, remote: SocketAddr) {
        self.state = HandshakeState::WaitingForResponse;
        self.remote_addr = Some(remote);
        self.started_at = Some(Instant::now());
    }

    pub fn timed_out(&self) -> bool {
        match self.started_at {
            Some(start) => start.elapsed() > self.timeout,
            None => true,
        }
    }
}

File: ports/python/rushaudio/handshake.py
"""Session handshake state machine and payload builders/parsers.

Mirrors ``src/session/handshake.rs`` in the Rust reference implementation and
spec section 5.
"""

import enum
import time
from dataclasses import dataclass, field
from typing import Optional, Tuple

from . import constants as _c
from .packet import Packet
from .types import AudioCodec, PacketType, StreamConfig


class HandshakeRole(enum.Enum):
    INITIATOR = "Initiator"
    RESPONDER = "Responder"


class HandshakeState(enum.Enum):
    IDLE = "Idle"
    WAITING_FOR_RESPONSE = "WaitingForResponse"
    WAITING_FOR_REQUEST = "WaitingForRequest"
    COMPLETED = "Completed"
    FAILED = "Failed"


@dataclass
class Handshake:
    """Tracks one side of a handshake and builds/parses its packets."""

    role: HandshakeRole
    ssrc: int
    config: StreamConfig
    state: HandshakeState = HandshakeState.IDLE
    remote_addr: Optional[str] = None
    started_at: Optional[float] = None
    timeout: float = _c.DEFAULT_HANDSHAKE_TIMEOUT_MS / 1000.0

    def build_request(self, seq: int, ts: int) -> Packet:
        """Build a HandshakeRequest packet (14-byte payload, spec 4.3)."""
        payload = (
            self.ssrc.to_bytes(4, "big")
            + self.config.sample_rate.to_bytes(4, "big")
            + self.config.bitrate.to_bytes(4, "big")
            + bytes((self.config.codec.to_u8(), self.config.channels))
        )
        return Packet.new(PacketType.HANDSHAKE_REQUEST, seq, ts, payload)

    def build_response(self, seq: int, ts: int, accepted: bool) -> Packet:
        """Build a HandshakeResponse packet (5-byte payload, spec 4.4)."""
        payload = bytes((0x01 if accepted else 0x00,)) + self.ssrc.to_bytes(4, "big")
        return Packet.new(PacketType.HANDSHAKE_RESPONSE, seq, ts, payload)

    @staticmethod
    def parse_request(packet: Packet) -> Optional[Tuple[int, StreamConfig]]:
        """Parse a HandshakeRequest -> (remote_ssrc, StreamConfig)."""
        payload = packet.payload
        if len(payload) < 12:
            return None

        remote_ssrc = int.from_bytes(payload[0:4], "big")
        sample_rate = int.from_bytes(payload[4:8], "big")
        bitrate = int.from_bytes(payload[8:12], "big")
        codec = AudioCodec.from_u8(payload[12]) if len(payload) > 12 else None
        if codec is None:
            codec = AudioCodec.OPUS
        channels = payload[13] if len(payload) > 13 else 2

        config = StreamConfig(
            sample_rate=sample_rate,
            channels=channels,
            bitrate=bitrate,
            codec=codec,
            ssrc=0,
        )
        return remote_ssrc, config

    @staticmethod
    def parse_response(packet: Packet) -> Optional[Tuple[bool, int]]:
        """Parse a HandshakeResponse -> (accepted, responder_ssrc)."""
        payload = packet.payload
        if len(payload) < 5:
            return None
        accepted = payload[0] == 0x01
        ssrc = int.from_bytes(payload[1:5], "big")
        return accepted, ssrc

    def start(self, remote: str) -> None:
        self.state = HandshakeState.WAITING_FOR_RESPONSE
        self.remote_addr = remote
        self.started_at = time.monotonic()

    def timed_out(self) -> bool:
        if self.started_at is None:
            return True
        return (time.monotonic() - self.started_at) > self.timeout